### PR TITLE
Add apps/resnet_50

### PR DESCRIPTION
**Pull request by asaadaldien**
Add a manually scheduled resnet50 implementation to apps/resnet_50.

```
$ make benchmark_and_validate
Running Resnet50 for 10 iterations....
Execution time : 438.491ms
Class for random data of seed 123 is 783
Writing output layer to ./bin/res50gen_output.bin
python3 validate_resnet50_output.py ./bin/res50gen_output.bin 123
Success!
```

### CODE REVIEW

File: apps/resnet_50/Resnet50Generator.cpp [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+            resunit_relu[block_id] = relu_layer(resunit_sum[block_id], "block" + std::to_string(block_id) + "_res_relu");
+
+            // create final 3 layers
+            if (block_id == 15) {
+                pool5 = avg_pool_layer(resunit_relu[block_id], pool5_ws, "pool5");
+                fc1000 = fc_layer(pool5, fc1000_ws, fc1000_weights, fc1000_bias, "fc");
+                final_output = softmax_layer(fc1000, 1000, "softmax");
+            }
+        }
+        conv1.f.compute_root();
```
Comment (abadams) [inline apps/resnet_50/Resnet50Generator.cpp]: There isn't really a schedule here yet, just enough of one to compile. I'm worried people will advertise that they are X times faster than Halide without realizing this. Please add:
```
// TODO: Actually schedule this. 
```

and where it prints the runtime in process.cpp please make it also print "This code hasn't been scheduled properly yet so this runtime isn't representative of anything and should not be used as a basis for any comparisons".

When I get around to actually writing a schedule I'll remove the prints.

Comment (asaadaldien) [reply apps/resnet_50/Resnet50Generator.cpp]: Here you go...
```
$ make benchmark_and_validate
bin/process 10  ./bin/pytorch_weights/ 123 ./bin/res50gen_output.bin
Running Resnet50 for 10 iterations....
*************************** Please note ******************************
This code hasn't been scheduled properly yet so this runtime
isn't representative of anything and should not be used as a basis
for any comparisons.
Execution time : 357.742ms
**********************************************************************
Class for random data of seed 123 is 783
Writing output layer to ./bin/res50gen_output.bin
python3 validate_resnet50_output.py ./bin/res50gen_output.bin 123
Success!
```

**Comment by steven-johnson** [comment]
What's the status on this -- ready to land?

File: apps/resnet_50/Makefile [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+$(BIN)/resnet50.a: $(BIN)/resnet50.generator
+	@mkdir -p $(@D)
+	$^ -g resnet50 -o $(BIN) -f resnet50 target=$(HL_TARGET) auto_schedule=false
+
+$(BIN)/process: process.cpp $(BIN)/resnet50.a 
+	@mkdir -p $(@D)
+	$(CXX) $(CXXFLAGS) -I$(BIN) -Wall -O3 $^ -o $@ $(LDFLAGS) $(IMAGE_IO_FLAGS) $(CUDA_LDFLAGS) $(OPENCL_LDFLAGS) $(OPENGL_LDFLAGS)
+
+benchmark_and_validate: $(BIN)/process $(BIN)/pytorch_weights/ok
```
Comment (abadams) [inline apps/resnet_50/Makefile]: We should add resnet50 to TEST_APPS in the top-level Makefile, and add a "test" target to this makefile that ensures the generator builds (i.e. just build bin/process).

It won't validate correctness (which would require all the pytorch deps, and downloading resnet weights), but we'll have this app as a compile-time canary for future PRs.

Comment (asaadaldien) [reply apps/resnet_50/Makefile]: Done! I added app/resnet_50 build to test_app in the last commit.

**Comment by abadams** [comment]
Confirmed that it's running the generator and passing on all buildbots. Should act as a good canary for anything related to compiling large pipelines as single translation units.

**Pull request merged**